### PR TITLE
1125 scheduler: template runInEpoch* functions 

### LINKED ISSUE #1125: runInEpoch* do not take a universal reference (they take a r-val ref)
**What Needs to be Done?**

Due to this, you can pass an existing std::function to them. Template them and move to the header file. Something like this would work:

```C++
template <typename Callable>
void runInEpochCollective(Callable&& fn) {
  //...
}
```

### PR DESCRIPTION
**Pull request by cz4rs**
Fixes #1125 

- [x] implementation
- [x] unit tests

### CODE REVIEW

File: src/vt/scheduler/scheduler.impl.h [hunk 1]
```diff
@@ -45,7 +45,39 @@
 #if !defined INCLUDED_VT_SCHEDULER_SCHEDULER_IMPL_H
 #define INCLUDED_VT_SCHEDULER_SCHEDULER_IMPL_H
 
+#include "vt/scheduler/scheduler.h"
```
Comment (PhilMiller) [inline src/vt/scheduler/scheduler.impl.h:48]: This is a circular inclusion, and should be unnecessary

Comment (cz4rs) [reply src/vt/scheduler/scheduler.impl.h]: removed :+1:

File: tests/unit/scheduler/test_high_level_calls.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#include <gtest/gtest.h>
+
+#include "vt/transport.h"
+#include "test_parallel_harness.h"
+
+namespace vt { namespace tests { namespace unit {
+
+struct TestHighLevelCalls : TestParallelHarness {
+  bool called = false;
+};
```
Comment (PhilMiller) [inline tests/unit/scheduler/test_high_level_calls.cc:54]: Just as a check, this gets instantiated separately for each of the tests below, correct?

Comment (PhilMiller) [reply tests/unit/scheduler/test_high_level_calls.cc:54]: Assuming that's affirmed, I'm +1 on this

File: src/vt/scheduler/scheduler.impl.h [hunk 3]
```diff
@@ -46,6 +46,36 @@
...
+void runInEpochRooted(Callable&& fn) {
+  theSched()->triggerEvent(sched::SchedulerEvent::PendingSchedulerLoop);
+
+  auto ep = theTerm()->makeEpochRooted();
+  theMsg()->pushEpoch(ep);
+  fn();
+  theMsg()->popEpoch(ep);
+  theTerm()->finishedEpoch(ep);
+  runSchedulerThrough(ep);
+}
```
Comment (jstrzebonski) [inline src/vt/scheduler/scheduler.impl.h:76]: Those two functions are almost identical. Maybe we would like to do something like this?
```
template <EpochType (term::TerminationDetector::*memFun)()>
struct runInEpoch {
  template <typename Callable>
  static void run(Callable&& fn) {
    theSched()->triggerEvent(sched::SchedulerEvent::PendingSchedulerLoop);

    auto ep = (theTerm()->*memFun)();
    theMsg()->pushEpoch(ep);
    fn();
    theMsg()->popEpoch(ep);
    theTerm()->finishedEpoch(ep);
    runSchedulerThrough(ep);
  }
};

template <typename Callable>
void runInEpochCollective(Callable&& fn) {
  runInEpoch<&term::TerminationDetector::makeEpochCollective>::run(
    std::forward<Callable>(fn)
  );
}

template <typename Callable>
void runInEpochRooted(Callable&& fn) {
  runInEpoch<&term::TerminationDetector::makeEpochRooted>::run(
    std::forward<Callable>(fn)
  );
}
```

File: tests/unit/scheduler/test_high_level_calls.cc (hunk 2 above)
Comment (cz4rs) [reply tests/unit/scheduler/test_high_level_calls.cc:54]: yes, it's a separate instance for each `TEST_F`

File: src/vt/scheduler/scheduler.impl.h (hunk 3 above)
Comment (cz4rs) [reply src/vt/scheduler/scheduler.impl.h:76]: I guess simply passing an epoch to a generalized function (like runInEpoch) would work? I will have a look

**Review by lifflander — APPROVED** [review]
This looks great to me

File: src/vt/scheduler/scheduler.impl.h [hunk 4]
```diff
@@ -46,6 +46,35 @@
...
+
+template <typename Callable>
+void runInEpochCollective(Callable&& fn) {
+  auto ep = theTerm()->makeEpochCollective();
+  runInEpoch(ep, fn);
```
Comment (jstrzebonski) [inline src/vt/scheduler/scheduler.impl.h:68]: I guess `runInEpoch(ep, std::forward(fn));?

**Review by jstrzebonski — APPROVED** [review]
Seems to be OK.

**Review by JacobDomagala — APPROVED** [review]

**Comment by cz4rs** [comment]
(re-running to get Codacy/PR Quality Review to pass)

**Pull request merged**